src/_gettsim/interface.py

```diff
@@ -2,661 +2,453 @@
-
-        # Referenced `p_id` must exist in the input data
-        if not data[foreign_key].isin(p_ids).all():
-            message = (
-                f"The following {foreign_key}s are not a valid p_id in the input data:"
-                f" {list(data[foreign_key].loc[~data[foreign_key].isin(p_ids)])}"
-            )
-            raise ValueError(message)
-
-        # Referenced `p_id` must not be the same as the `p_id` of the same row
-        if (data[foreign_key] == data["p_id"]).any():
-            message = (
-                f"The following {foreign_key}s are equal to the p_id in the same row:"
-                f" {list(data[foreign_key].loc[data[foreign_key] == data['p_id']])}"
-            )
-            raise ValueError(message)
 
+    # Create root nodes tree
+    root_nodes_view = nx.subgraph_view(dag, filter_node=lambda n: dag.in_degree(n) == 0)
+    root_nodes_tree = tree_structure_from_paths(
+        [tuple(node.split(QUALIFIED_NAME_SEPARATOR)) for node in root_nodes_view.nodes]
+    )
 
-def _fail_if_root_nodes_are_missing(root_nodes, data, functions):
-    # Identify functions that are part of the DAG, but do not depend
-    # on any other function
-    funcs_based_on_params_only = [
-        func_name
-        for func_name, func in functions.items()
-        if len(
-            [a for a in inspect.signature(func).parameters if not a.endswith("_params")]
-        )
-        == 0
-    ]
-
-    missing_nodes = [
-        c for c in root_nodes if c not in data and c not in funcs_based_on_params_only
-    ]
-    if missing_nodes:
-        formatted = format_list_linewise(missing_nodes)
-        raise ValueError(f"The following data columns are missing.\n{formatted}")
-
+    _fail_if_root_nodes_are_missing(
+        functions_tree=functions_tree,
+        data_tree=data_tree,
+        root_nodes_tree=root_nodes_tree,
+    )
 
-def _reduce_to_necessary_data(root_nodes, data, check_minimal_specification):
-    # Produce warning or fail if more than necessary data is given.
-    unnecessary_data = set(data) - root_nodes
-    formatted = format_list_linewise(unnecessary_data)
-    message = f"The following columns in 'data' are unused.\n\n{formatted}"
-    if unnecessary_data and check_minimal_specification == "warn":
-        warnings.warn(message, stacklevel=2)
-    elif unnecessary_data and check_minimal_specification == "raise":
-        raise ValueError(message)
+    # Get only part of the data tree that is needed
+    input_data_tree = partition_tree_by_reference_tree(
+        tree_to_partition=data_tree,
+        reference_tree=root_nodes_tree,
+    )[0]
 
-    return {k: v for k, v in data.items() if k not in unnecessary_data}
+    # Convert to numpy.ndarray
+    return optree.tree_map(lambda x: x.to_numpy(), input_data_tree)
 
 
-def _round_and_partial_parameters_to_functions(functions, params, rounding):
-    """Create a dictionary of all functions that are available.
+def _round_and_partial_parameters_to_functions(
```

Comment:
Should just be 
```suggestion
def _partial_parameters_to_functions(
```
the line doing the rounding can be done in the calling code.

Feel free to postpone to a different PR.